[WIP][SQL][SPARK-6632]: Read schema from each input split in the ReadSupport hook, reconciling with the metastore schema at that time

saucam opened the pull request:
Hey @liancheng,

How about this approach for schema reconciliation, where we use the metastore schema, and reconcile within the ReadSupport init function. This way, we handle each input file in the map task, and no need to read schema from all part files and merging before initiating the tasks.
I have not removed the schema merging code/ test cases for now. Let me know your thoughts on this one.

Commit:
SPARK-6632: Read schema from each input split in the ReadSupport hook, reconciling with the metastore schema at that time

AmplabJenkins commented:
Can one of the admins verify this patch?

liancheng commented:
ok to test

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29714/
Test FAILed.

Commit:
SPARK-6632: Fix whitespace

liancheng commented:
Ah, I'm also considering similar optimizations for Spark 1.4 :)

The tricky part here is that, when scanning the Parquet table, Spark needs to call `ParquetInputFormat.getSplits` to compute (Spark) partition information. This `getSplits` call can be super expensive as it needs to read footers of all Parquet part-files to compute the Parquet splits. And that's why `ParquetRelation2` caches those footers at the very beginning and inject them into an extended Parquet input format. With all these footers cached, `ParquetRelation2.readSchma()` is actually quite lightweight. So the real bottleneck is reading all those footers.

Fortunately, Parquet is also trying to avoid reading footers entirely at the driver side (see https://github.com/apache/incubator-parquet-mr/pull/91 and https://github.com/apache/incubator-parquet-mr/pull/45). After upgrading to Parquet 1.6, which is expected to be released next week, we can do this properly for better performance.

So ideally, we don't read footers on driver side, and when we have a central arbitrative schema at hand, either from metastore or data source DDL, we don't do schema merging at driver side either. I haven't got time to walk through all related Parquet code path and PRs yet, so the above statements may be inaccurate. Please correct me if you find any mistakes.

saucam commented:
hmm i see. Would definitely go through these PRs. Anyways fixed the whitespace problem here.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29716/
Test FAILed.

liancheng commented:
add to whitelist

SparkQA commented:
[Test build #29958 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29958/consoleFull) for   PR 5298 at commit [`89efac5`](https://github.com/apache/spark/commit/89efac5a8392807a422cfb7826b9458c285335c6).

SparkQA commented:
[Test build #29958 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29958/consoleFull) for   PR 5298 at commit [`89efac5`](https://github.com/apache/spark/commit/89efac5a8392807a422cfb7826b9458c285335c6).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29958/
Test FAILed.

Commit:
SPARK-6632: maybeMetastoreSchema should only contain non-partitioning columns, fullschema is derived later within ParquetRelation2
            This is done so that partitionKeysIncludedInParquetSchema is computed correctly later on

Commit:
SPARK-6632: Fix whitespace

SparkQA commented:
[Test build #30118 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30118/consoleFull) for   PR 5298 at commit [`866aa93`](https://github.com/apache/spark/commit/866aa938ef5b14ef743af739797791d08aaac04d).

SparkQA commented:
[Test build #30120 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30120/consoleFull) for   PR 5298 at commit [`90d7782`](https://github.com/apache/spark/commit/90d77825922e6ac29051febd60cb8a9ed1b9841f).

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30119/
Test FAILed.

SparkQA commented:
[Test build #30120 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30120/consoleFull) for   PR 5298 at commit [`90d7782`](https://github.com/apache/spark/commit/90d77825922e6ac29051febd60cb8a9ed1b9841f).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30120/
Test PASSed.

SparkQA commented:
[Test build #30118 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30118/consoleFull) for   PR 5298 at commit [`866aa93`](https://github.com/apache/spark/commit/866aa938ef5b14ef743af739797791d08aaac04d).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30118/
Test PASSed.

saucam commented:
hey @liancheng , this change now reconciles schema within the tasks. do suggest. After that I will remove the merge schema functions that are no longer needed

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTableSupport.scala:
```diff
@@ -98,12 +98,32 @@ private[parquet] class RowReadSupport extends ReadSupport[Row] with Logging {
+          throw new RuntimeException(s"""Field ${attr.name} is non-nullable, 
+            but not found in parquet file schema: ${fileSchema}""".stripMargin)
+        }}})
+
```
Are we essentially duplicating `ParquetRelation2.mergeMetastoreParquetSchema` here?

saucam replied:
yes, the difference being that this happens within each task, whereas ParquetRelation2.mergeMetastoreParquetSchema happens on the driver. This eliminates the need of mergeMetastoreParquetSchema method

liancheng replied:
`ParquetRelation2.mergeMetastoreParquetSchema` is just a static method, can we just reuse that here? Especially comments for this method and `ParquetRelation2.mergeMissingNullableFields` are pretty useful. I would like to keep them.

And please don't put multiple `}`/`)` on a single line.

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTableSupport.scala:
```diff
@@ -98,12 +98,32 @@ private[parquet] class RowReadSupport extends ReadSupport[Row] with Logging {
     val requestedAttributes = RowReadSupport.getRequestedSchema(configuration)
 
+    // convert fileSchema to attributes
+    val fileAttributes = ParquetTypesConverter.convertToAttributes(fileSchema, true, true)
```
I'm afraid we can't always set the last two arguments to `true`, they should be determined according to corresponding `SQLConf` configurations.

saucam replied:
these booleans are for finding the datatype of the attribute, whereas here we are just interested in finding out the names of the columns, to reconcile with metastore schema. Hence it is safe to always send these parameters as true, since we do not have SQL context here from which to derive these.

liancheng replied:
I see, thanks for the explanation. Would please also add a comment for this?

liancheng commented on sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTableSupport.scala:
```diff
@@ -98,12 +98,32 @@ private[parquet] class RowReadSupport extends ReadSupport[Row] with Logging {
+          // this is just a safety check since in hive all columns are nullable
+          // throw exception here
+          throw new RuntimeException(s"""Field ${attr.name} is non-nullable, 
+            but not found in parquet file schema: ${fileSchema}""".stripMargin)
```
This exception message would look pretty weird when printed since there isn't a "margin" character (`|`) in the string. Also we don't want to split this message into multiple lines. Maybe this:

```scala
throw new RuntimeException(
  s"Field ${attr.name} is non-nullable, " +
    s"but not found in parquet file schema: ${fileSchema}")
```

liancheng commented:
@saucam Right now I feel kinda hesitant to have this. As explained in my previous comment, the major bottleneck for Parquet metadata handling happens when reading footers. Without eliminating this, moving schema merging to task side doesn't bring performance benefits (although I haven't done any benchmark for this PR yet). Plus, there are risks of introducing regressions.

However, this PR is still very valuable as it proves this approach is doable. Eventually, we would like to have this after upgrading Parquet to 1.6.0 and add the ability to avoid reading footers on driver side whenever a global arbitrative schema is available. I've opened [SPARK-6795] [1] to track this issue. I will probably start working on SPARK-6795 later this month. Would you mind me revisiting this at that time?

[1]: https://issues.apache.org/jira/browse/SPARK-6795

saucam commented:
ok @liancheng 
Thanks for the comments. In the meantime let me try to address your suggestions. Can we keep this open in WIP state for now ?

liancheng commented:
Yeah, sure.

marmbrus commented:
Hi @saucam, thanks for working on this.  I think that a lot of this has been implemented in Spark 1.5.  Can we close this issue?
